src/d3a/models/myco_matcher/myco_external_matcher.py

```diff
@@ -162,45 +171,69 @@ def _get_bids_offers(market: TwoSidedMarket, filters: Dict):
+        cls.validate_orders_exist_in_market(matcher, recommendation)
+
+    @classmethod
+    def validate_and_report(cls, matcher: MycoExternalMatcher, recommendations: Dict) -> Dict:
+        """Validate recommendations and return a detailed report."""
+
+        response = {"status": "success", "recommendations": []}
         for recommendation in recommendations:
-            if not BidOfferMatch.is_valid_dict(recommendation):
-                raise MycoValidationException("BidOfferMatch is not valid")
-            market = self.markets_mapping.get(recommendation.get("market_id"))
-            if market is None:
-                # The market doesn't exist
-                raise MycoValidationException(
-                    f"Market with id {recommendation.get('market_id')} doesn't exist.")
-            if market.readonly:
-                # The market is already finished
-                # TODO: we're clearing markets cache after each market cycle so is this check
-                # TODO: .. really relevant? (All finished markets should be deleted already)
-                raise MycoValidationException(
-                    "Cannot match trades in a finished market.")
-
-            # Get the original bid and offer from the market
-            market_bids = [
-                market.bids.get(bid.get("id")) for bid in recommendation.get("bids")]
-            market_offers = [
-                market.offers.get(offer.get("id")) for offer in recommendation.get("offers")]
-
-            if not (all(market_bids) and all(market_offers)):
-                # Offers or Bids either don't belong to market or were already matched
-                continue
             try:
-                market.validate_bid_offer_match(
-                    market_bids,
-                    market_offers,
-                    recommendation.get("trade_rate"),
-                    recommendation.get("selected_energy")
+                cls._validate(matcher, recommendation)
+                response["recommendations"].append(
+                    {**recommendation, "status": "success"}
                 )
-            except InvalidBidOfferPairException:
-                continue
-
-            recommendation["bids"] = [bid.serializable_dict() for bid in market_bids]
-            recommendation["offers"] = [offer.serializable_dict() for offer in market_offers]
-            validated_recommendations.append(recommendation)
-        return validated_recommendations
+            except Exception as exception:
+                if isinstance(exception, cls.BLOCKING_EXCEPTIONS):
+                    response["status"] = "fail"
+                    response[
+                        "message"] = f"Validation Error, matching will be skipped: {exception}"
+                    break
+                else:
+                    response["recommendations"].append(
+                        {**recommendation, "status": "fail", "message": str(exception)})
+                    continue
```

Comment:
Nice! Perfect example for motivation of not using simple `Exception`s, right @fievelk ? This is what you meant in Lisbon, right?

Reply:
Update, I have removed the broad `except Exception` to not report **any** exception to the client

Reply:
@hannesdiedrich this can be one of such cases where you want to have a distinction between different types of exceptions, yes!